src/norm/model.nim

```diff
@@ -114,3 +114,25 @@ proc checkRo*(T: typedesc[Model]) =
   when T.hasCustomPragma(ro):
     {.error: "can't use mutating procs with read-only models".}
 
+proc getRelatedFieldNameOn*[T: Model, M: Model](targetModel: typedesc[T], sourceModel: typedesc[M]): string  {.raises: [FieldDefect].} =
```

Comment:
> It takes two types, T and M. We assume M is an object type with a field of type T. The function returns the name of the field in M that has the type T.

Almost, it's capabilities are in fact a bit broader than that. The call `getRelatedFieldNameTo(S, T)`  returns the name of the field in S that "points to the same table that model T has". Which means this also works if S has a field with an fk-pragma that points to T. And it **also** works if S has a field of typeT2, if T2 has **the same tablename** as T via the `tableName` pragma. 

1. Using objects will not be possible here as the proc acts at compile time and I'm not aware how you would have objects at compile time for analysis. I'm personally insistent on keeping all of these calculations at compileTime, because the earlier you see these issues the better. If I can avoid a user running into an issue at runtime by having their code not compile when they do something wrong, I'd rather do that with a nice error message than let them notice at runtime. 
Regarding Datetime: Fascinatingly, it doesn't pose a problem. I assumed the same but tried it out in case it worked and found that DateTime does not cause issues here! Likely because this all acts at compileTime. I'll attach a code example to the bottom of this comment so you can run it and see it yourself. Here a [link to that code in play nim lang](https://play.nim-lang.org/#ix=3Q9n). 
2. You get the first field that matches, exactly as you assumed. As such I changed things around a bit. The core proc `selectOneToMany` now takes an additional parameter `foreignKeyFieldName` to circumvent *exactly* the problem you lined out there. This basically makes this proc almost obsolete. However, I added an overload of `selectOneToMany` that is without the parameter `foreignKeyFieldName` parameter and instead uses `getRelatedFieldNameTo` to infer it. You can still run into this scenario of 2 foreign key fields to the same model, however you'll get a FieldDefect at compile time with a nice error message if you do. See the test `"When there is multiple many-to-one relationships between two models and the type field for fetching the desired relationship is not specified, then do not compile"` (postgres/trelated.nim line 88).
3. Sadly not possible, doing so will cause the code to not compile, as the proc does have side-effects. I think the compiler counts the FieldDefects that might be raised as such side-effects.
4. I agree that the naming isn't great. `fieldsOfModelType` would be inaccurate though, as you only get a single foreign-key-field on the source model pointing to the target model. Strictly speaking even my name is accurate, since the model-field on `source` and the model specified in `target` don't even have to match, they merely need to point to the same table. Of course I could change that, but that feels like narrowing the capabilities of this proc down unnecessarily. Anyway, I agree that the name is bad, I am however also not satisfied with `fieldsOfModeltype`. If I name it how I use it I guess it would be `inferRelatedFieldNameToModel` or something like it, but I'm not convinced of that name either.


```nim
import norm/[model, pragmas]
import std/[typetraits, strformat, options, times, macros]


proc getRelatedFieldNameTo*[M: Model](targetTableName: static string, normModel: typedesc[M]): string {.compileTime.} =
  ## A compile time proc that searches the given `normModel` type for any 
  ## foreign key field that points to a table with the given name `targetTableName`. 
  ## Raises a FieldDefect at compile time if the model does not have exactly 
  ## one foreign key field to that table.
  var fieldNames: seq[string] = @[]
  const name = typetraits.name #Allows this generic proc to always have typetraits.name proc available even when the context it is called from doesn't import typetraits
  
  for sourceFieldName, sourceFieldValue in M()[].fieldPairs:
      #Handles case where field is an int64 with fk pragma
      when sourceFieldValue.hasCustomPragma(fk):
        when targetTableName == sourceFieldValue.getCustomPragmaVal(fk).table():
          fieldNames.add(sourceFieldName)
      
      #Handles case where field is a Model type
      elif sourceFieldValue is Model:
        when targetTableName == sourceFieldValue.type().table():
          fieldNames.add(sourceFieldName)
      
      #Handles case where field is a Option[Model] type
      elif sourceFieldValue is Option:
        when sourceFieldValue.get() is Model:
          when targetTableName == genericParams(sourceFieldValue.type()).get(0).table():
              fieldNames.add(sourceFieldName)

  if fieldNames.len() == 1:
    return fieldNames[0]
  
  elif fieldnames.len() < 1:
    let errorMsg = fmt "Tried getting foreign key field from model '{name(M)}' to model '{targetTableName}' but there is no such field!"
    raise newException(FieldDefect, errorMsg)
  
  elif fieldnames.len() > 1:
    let errorMsg = fmt "Can't infer foreign key field from model '{name(M)}' to model '{targetTableName}'! There is more than one foreign key field to that table!"
    raise newException(FieldDefect, errorMsg)

proc getRelatedFieldNameTo*[S: Model, T:Model](source: typedesc[S], target: typedesc[T]): string {.compileTime.} =
  ## A compile time proc that analyzes the given `source` model for any foreign key field that points to
  ## the table of the `target` model. Raises a FieldDefect at compile time if the model does not
  ## have exactly one foreign key field to that table 
  result = getRelatedFieldNameTo(T.table(), source)

type A* = ref object of Model
  time*: DateTime

type B* = ref object of Model
  somename*: string
  time*: DateTime
  fkToA*: A

type C* = ref object of Model
  da*: string
  time*: DateTime
  fkToA* {.fk: A.}: int64

type D* = ref object of Model
  du*: string
  time*: DateTime
  fkToA*: Option[A]


const val1 = B.getRelatedFieldNameTo(A)
const val2 = C.getRelatedFieldNameTo(A)
const val3 = D.getRelatedFieldNameTo(A)
echo val1
echo val2
echo val3
```